treewide: Basic Ubuntu Touch & Halium enablement

fredldotme opened the pull request:
This PR contains:

- Release's VARIANT_ID being detected as "touch" variant of Ubuntu
- Enabling extrausers for that
- Enabling Docker to work on Ubuntu Touch by avoiding problematic Docker AppArmor insertions
- snap-confine working to set up /android and required symlinks pointing to Halium-side libraries

For more information about Halium: https://halium.org

Diverged greatly from the NVIDIA support code it's based on and provides a working environment for confined Qt5, Qt6, SDL, GTK and Flutter Snaps (as long as a GLES renderer is available) like QML Creator (`sudo snap install qmlcreator`) with the `hybris-2404` GNU/Linux drivers.

SNAPDENG-36010

bboozzoo commented on release/release.go:
```diff
@@ -187,6 +187,11 @@ var OnClassic bool
+// OnTouch states whether the process is running inside a Touch image
+// with a classic (but mostly read-only) filesystem layout.
+// It is a narrowed down variant of OnClassic
+var OnTouch bool
```
I think I'd split the part specific to Touch into a separate PR. There may be questions about how we want to approach this specific variant that are unrelated to supporting libhybris

fredldotme replied:
Tbf I've waited pretty long on the previous PR that has been sitting for a while, which used to be the libhybris-only PR. But if more people want to have it split up, then sure.

bboozzoo replied:
I would split it. The s-c bits should be fairly easy to land, but folks may be asking questions about this one and there's no point in holding back the rest of the PR.

fredldotme replied:
Done.

bboozzoo reviewed (commented):
Thank you. Please have a look at the comments.

fredldotme commented on cmd/snap-confine/mount-support-hybris.c:
```diff
@@ -0,0 +1,371 @@
+//
+// The glob list passed to us is produced with paths relative to source dir,
+// to simplify the various tie-in points with this function.
+static void sc_hybris_populate_libgl_with_hostfs_symlinks(const char *libgl_dir,
```
Sure, will do

fredldotme replied:
Taken care of.

fredldotme commented on cmd/snap-confine/mount-support-hybris.c:
```diff
@@ -0,0 +1,371 @@
+	}
+}
+
+static void sc_hybris_mkdir_and_mount_and_glob_files(const char *rootfs_dir,
```
Yup, noted

fredldotme replied:
Done.

fredldotme commented on cmd/snap-confine/snap-confine.c:
```diff
@@ -472,6 +472,14 @@ int main(int argc, char **argv)
+        // variables preparing hybris support correctly. If GL is not permitted the
+        // variables won't be of much use, but still set them up here.
+        if (!invocation.classic_confinement && access("/system/build.prop", F_OK) == 0) {
+            setenv("HYBRIS_LINKER_DIR", "/var/lib/snapd/lib/gl/libhybris/linker", 1);
+            setenv("HYBRIS_EGLPLATFORM_DIR", "/var/lib/snapd/lib/gl/libhybris", 1);
```
IMO having snap-confine set up the environment by itself allows for:

- Expected environment for any Snap
- Less ways of the Snap screwing up something in the setup (because it misses -desktop-integration or does things itself)
- Have existing Snaps working as-is.

I cannot expect upstream app developers to care or have the knowledge to setup libhybris appropriately, which is why I put it there.

bboozzoo replied:
Hm I think I'd rather place it somewhere around here: https://github.com/snapcore/snapd/blob/68fe7eb9f7496ec54f2b3373a5c8d1993b9eb822/cmd/snap-exec/main.go#L211-L213 there's a precedent for it already for Cups

cc @zyga ?

fredldotme replied:
Taken care of.

fredldotme commented:
Since security is supposed to be involved in the discussion, here is a little hint about something that might come up:

For Halium/Android 9-only devices we had to ship a TLS-padding hack in the form of a shared library that needs to be LD_PRELOADed into the process, otherwise bionic libraries and glibc libraries would have a mismatch in TLS sizes. The do-nothing library resides here: https://gitlab.com/ubports/development/core/hybris-support/tls-padding/-/blob/main/tls-padding.cpp?ref_type=heads

Ubuntu Touch sets the LD_PRELOAD environment variable to include this hack into every user-app process, so ideally we can keep it to keep having 9 devices chugging along.

It explains the necessity to include the tls-padding .so into the AppArmor profile so that it can be mapped into the process image.

fredldotme commented on cmd/snap-confine/mount-support-hybris.c:
```diff
@@ -0,0 +1,371 @@
+static void sc_hybris_mount_android_rootfs(const char *rootfs_dir)
+{
+	// Bind mount a tmpfs on $rootfs_dir/$tgt_dir (i.e. /var/lib/snapd/lib/gl)
+	char rfsbuf[512] = { 0 };
```
Addressed.

fredldotme commented:
Thanks for the additional comments, will take care of them.

I just noticed I wasn't specific enough on the reasoning regarding commit https://github.com/snapcore/snapd/pull/13661/commits/d779cae7a71112297698b6d438e636acccafe051 where I stated "on a separate filesystem". This means udev isn't able to manage it using devtmpfs.

On a sidenote I have signed the CLA on Launchpad for Unity 7 and other projects so I was wondering what the status of validity is.

zyga commented:
I will happily review this after 2.62 is released. Apologies for the delay, we're really busy trying to get the release in good shape.

zyga commented on interfaces/builtin/opengl.go:
```diff
@@ -183,6 +183,48 @@ unix (bind,listen) type=seqpacket addr="@cuda-uvmfd-[0-9a-f]*",
+/{,android/}apex/com.android.i18n/lib{,64}/**.so     m,
+/{,dev/}socket/property_service rw, # attach_disconnected path
+/{,dev/}socket/logdw rw, # attach_disconnected path
+/{,dev/}__properties__/** rw, # attach_disconnected path
```
Curious, where is __properties__ coming from? Is this some Android-specific interface? What's inside?

fredldotme replied:
The `__properties__` part contains SELinux info of process labels allowed to access Android properties through `getprop` and `setprop`. It's more of a void on UT though since functionality tied to SELinux is stubbed out to not conflict with AppArmor.

bboozzoo replied:
Does it need `rw` access or just read?

fredldotme replied:
Potentially only needs read-access, since that is set up in the Halium container. But I'll have to check.

fredldotme replied:
Addressed

fredldotme commented on interfaces/builtin/opengl.go:
```diff
@@ -183,6 +183,48 @@ unix (bind,listen) type=seqpacket addr="@cuda-uvmfd-[0-9a-f]*",
+/{,var/}run/shm/hybris_shm_data rw, # FIXME: LP: #1226569 (make app-specific)
+# This is a LD_PRELOADed TLS padding library allowing use of both
+# Android 9 and glibc TLS slots at the same time. Only used on Halium 9.
+/usr/lib/@{multiarch}/libtls-padding.so  mr,
```
This particular library resides on the host, which then gets picked up by snap-confine AFAICT.

In general thread-locals in a shared library are taken care of after initialization by `ld`, before the actual executable is run. Think calling constructors of a static variable, it's similar in style to how `ld` treats thread-locals.

bboozzoo replied:
Just for the record, host files do not appear at that location. What you see here comes from the base snap.

fredldotme replied:
I am surprised then that snap-confine correctly stops complaining about not being able to map that library.

bboozzoo replied:
If you're setting LD_PRELOAD in your environment, than the denial from s-c should be raised regardless of this interface, as s-c is running under a different profile. Now, setting LD_PRELOAD in the environment will not have an effect of the snap as IIRC LD_PRELOAD isn't carried over to the snap's environment. Now, if OTOH is being set within the snap, then yes, if you allow it explicitly in the file like here, the denial will go away even if the file doesn't exist. Can you paste the denial you saw?

fredldotme replied:
LD_PRELOAD is set in the host environment, yes. This results in two things:

- `ld` complaining during the preload itself: `ERROR: ld.so: object 'libtls-padding.so' from LD_PRELOAD cannot be preloaded (failed to map segment from shared object): ignored.`
- AppArmor giving the reason why:

```
[23609.467580] audit: type=1400 audit(1712049539.906:1289): apparmor="DENIED" operation="file_mmap" profile="/usr/lib/snapd/snap-confine" name="/usr/lib/aarch64-linux-gnu/libtls-padding.so" pid=33306 comm="snap-confine" requested_mask="m" denied_mask="m" fsuid=0 ouid=0
```

Of course the env var is set system-wide because we already need it on the host, but the confined snap also needs to inherit the padded TLS area.

bboozzoo replied:
The denial for snap-confine is expected, since the PR does not modify the profile, but since it's functioning I'm not sure I'd update it. 
Within the snap, unless something sets LD_PRELOAD again, that environment variable will be empty, because ld.so will strip it out before switching to snap-confine. Are you setting it explicitly in the snaps you use for testing?

fredldotme replied:
I am not setting them within the Snap's environment and the the process' file mappings don't seem to reference the tls-padding library as well. I wonder if somehow the padded TLS area might be set up already in snap-confine and kept across exec()s, but I honestly doubt that.

bboozzoo replied:
You can start a process in some test snap and then on the host either list its open files or inspect its memory map.

fredldotme replied:
That's what I did to no avail. Seems though my historical assumptions were wrong as the padding hack was a requirement for Halium/Android 10, not 9, and I don't have a H10 MediaTek device available at the moment.

On the other hand, could it be that GCC has also adapted to Android's TLS slot layout? In that case core20+ Snaps wouldn't necessarily need it. More infos here: https://reviews.llvm.org/D53906

fredldotme replied:
After giving it some time, I've come to the conclusion that we might still want to pass through the preload hack to the confined environment in case it's running on Touch specifically. Would that be feasible from your perspective?

bboozzoo replied:
ld.so will strip out LD_PRELOAD before it switches execution to snap-confine, i.e. snap-confine will not even see that environment variable, so there's no chance of passing it to the snap. Perhaps you can load it from /etc/environment when executing in the snap, but it would not be automatic.

fredldotme replied:
I was more thinking of measuring whether it's running on Touch via other means (checking for existance of a Halium-only file) and recreating the environment variable. It might sound hacky at first but there really seems to be no easy other way.

bboozzoo replied:
I presume you've ran some apps with this change built locally. Were they able to start and function properly?

fredldotme replied:
No I couldn't as my assumption about Halium 9 vs 10 was wrong. I thought one of my test devices was susceptible to the issue, whereas it was just Halium 10 MediaTek devices specifically (which seem to have TLS slots hardcoded in their bionic-based GL libs).

fredldotme commented:
> I will happily review this after 2.62 is released. Apologies for the delay, we're really busy trying to get the release in good shape.

Ping! I've noticed 2.62 landing, and I don't want to be left behind. :)

fredldotme commented:
For the record and completeness sake, here is the list of devices tested with these changes:

- Halium 7.1: Sony Xperia X (suzu)
- Halium 9.0: Google Pixel 3a (sargo)
- Halium 10: JingPad (jingpad_a1)
- Halium 11: Fairphone 4 (fp4) & Volla Phone 22 (mimameid)

fredldotme commented:
Sooo... anything new?

bboozzoo commented:
@alexmurray can you have a look?

alexmurray commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -42,6 +42,8 @@
     /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libseccomp.so* mr,
     /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libcap.so* mr,
+    # TLS padding hack for Android 9
+    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libtls-padding.so mr,
```
It seems like this should be removed given the prior discussions that it appears unneeded.

alexmurray commented on cmd/snap-confine/udev-support.c:
```diff
@@ -116,6 +116,34 @@ static void sc_udev_allow_nvidia(sc_device_cgroup *cgroup)
+static void sc_udev_allow_hybris(sc_device_cgroup *cgroup)
+{
+	struct stat sbuf;
+
```
nitpick: would this be easier to read if it was written as something like the following (note untested):

```c
const char *paths[] = {"/dev/binderfs/binder", "/dev/binderfs/hwbinder", "/dev/binder", "/dev/hwbinder"};
for (int i = 0; i < sizeof(paths)/sizeof(paths[0]); i++) {
    struct stat sbuf;
    if (stat(paths[i], &sbuf) == 0) {
        sc_device_cgroup_allow(cgroup, S_IFCHR, major(sbuf.st_rdev),
				       minor(sbuf.st_rdev));
    }
}
```

fredldotme replied:
Thanks, based it on your template. Handled in d5a85d6

alexmurray commented on interfaces/builtin/opengl.go:
```diff
@@ -183,6 +183,48 @@ unix (bind,listen) type=seqpacket addr="@cuda-uvmfd-[0-9a-f]*",
+/{,var/}run/shm/hybris_shm_data rw, # FIXME: LP: #1226569 (make app-specific)
+# This is a LD_PRELOADed TLS padding library allowing use of both
+# Android 9 and glibc TLS slots at the same time. Only used on Halium 9.
+/var/lib/snapd/hostfs/usr/lib/@{multiarch}/libtls-padding.so  mr,
```
Again this should be removed since it appears unnecessary

fredldotme replied:
Please check and address my previous comment above, we still need a solution for Mediatek Android 10 devices and I doubt waiting for much longer is a good idea.

fredldotme replied:
Alright, short-term I am fine without supporting MTK-10 devices for now, but I would really like to have this handled later.

Handled in d5a85d6

alexmurray reviewed (changes requested):
LGTM other than the LD_PRELOAD hack which from the existing comments seems unnecessary / unworkable via snap-confine and should be removed.

zyga commented on cmd/snap-confine/mount-support-hybris.c:
```diff
@@ -0,0 +1,198 @@
+	}
+	(void)sc_set_effective_identity(old);
+
+	(void)mount(SC_HYBRIS_ROOTFS, android_rootfs_dir, NULL, MS_BIND | MS_REC | MS_RDONLY, NULL);
```
Should this be checked and should s-c die if it does not work? If it is legitimately allowed to fail, as indicated by the `(void)` up front, can you add a comment just above
to explain why.

fredldotme replied:
Can take a deeper look later after `$DAYJOB`

fredldotme replied:
Handled in d5a85d6

zyga commented on cmd/snap-confine/udev-support.c:
```diff
@@ -116,6 +116,34 @@ static void sc_udev_allow_nvidia(sc_device_cgroup *cgroup)
+		sc_device_cgroup_allow(cgroup, S_IFCHR, major(sbuf.st_rdev),
+				       minor(sbuf.st_rdev));
+	}
+	if (stat("/dev/binder", &sbuf) == 0) {
```
Since this runs everywhere, what happens if someone installs, say, anbox and gets /dev/binder. Now all snaps will allow access to the binder. I don't have an immediate idea on how to make this better without some host-level hint.

fredldotme replied:
The host level hint can again be a check for existance of /system/build.prop, which under Halium distributions always has to exist.

fredldotme replied:
Handled in d5a85d6

zyga reviewed (commented):
I had a look from top to bottom. Some comments.

I would like to land the move to nvidia/mount logic separately to make the diff shorter. I can propose that separately.

fredldotme commented on cmd/snap-confine/mount-support-hybris.c:
```diff
@@ -0,0 +1,198 @@
+	sc_identity old = sc_set_effective_identity(sc_root_group_identity());
+	int res = mkdir(android_rootfs_dir, 0755);
+	if (res != 0 && errno != EEXIST) {
+		die("cannot create tmpfs target %s", android_rootfs_dir);
```
Handled in d5a85d6

fredldotme commented:
I've addressed all mentioned requested changes, but I came up with another one which I'll happily add in case I get a "makes sense" from you:

How about only adding those libhybris-related AppArmor changes to the template in case it detects `/system/build.prop` on the host? Should also reduce potential issues and makes clear the separation between mainline requirements + adding those for libhybris on top.

fredldotme commented:
Look, I know you have priorities so I've come up with a way we could ship the now-diverging patches from this PR in something collecting all Ubuntu Touch enablement work.

Right now we have Snapz0r at 1.5.0 and an upcoming 1.6.0 on top of latest git. So that means it should apply, but it's now intertwined with the extrausers enablement for Touch, so really I don't know how to proceed with you guys here. It all should work, but there are some things you might not agree with having supported on a default Ubuntu configuration outside of Touch. Hence now relying on `release.OnTouch` in this branch here too.

This merged-together relying-on-it branch works, it is tested and it will have users, as have the other releases of Snapz0r, and it shouldn't collide with regular Ubuntu environments, so that should all be covered.

Yes it also by now has changes for Docker to work on Touch too, with behavior only triggering when not OnTouch. So we have another diverging behavior which might be required no matter what.

I want this ball to get rolling, so what's your opinion?

fredldotme commented:
I would like to re-introduce this PR as all the culminative work that resulted in snapd working with all expectations set for a libhybris environment. Some individual apps can benefit from additional environment variables being set but that's out of scope for this Pull Request.

So here we have:
- Release's VARIANT_ID being detected
- Enabling extrausers for that
- Enabling most of Go-side Ubuntu Touch decisions
- Enabling Docker to work on Ubuntu Touch by avoiding problematic Docker AppArmor insertions
- snap-confine working to set up /android and required symlinks pointing to Halium-side libraries

Please consider merging this as one merge rather than splitting this up even further. Accepting the merge request would be a huge benefit to snapd users on Ubuntu Touch.

fredldotme commented:
Not to nag too much, just saying the list of changes coming up might pile up. snapd really needs love on Ubuntu Touch.

Regardless of that, changes are self-contained and if requested to split this up further I can just cherry-pick. But you need to tell me how you want it.

fredldotme commented:
My official statement to the recent rebase: I hope it captures everything that is mergeable from my PoV, it should compare to recent master and provide all basic blocks to enable hybris distros like Ubuntu Touch to run properly. Yes it adds `/android` into the environment if it finds the typical `/system/build.prop` but running GL requires those contents regardless.

fredldotme committed:
cmd/snap-confine: Initial support for mounting Halium libraries

The way that graphics support works with Snaps on Ubuntu Touch, we need snap-confine to
mount /android into the Snap's environment and set up compatibility symlinks. Then a
gpu content snap like hybris-2404 plugs the gpu-2404 interface with libhybris shipped
inside of the content snap. libhybris' EGL gets loaded and resolves libraries stored
in /android/{system,vendor,odm,apex}/lib{,64}, which results in a working graphics stack.

Determine this by checking for the existance of some Halium-typical paths and only proceed
if the running system matches a Halium one.

fredldotme committed:
release: Introduce OnTouch property for detecting Ubuntu Touch

Ubuntu Touch differs from typical classic distributions in that it is mounted read-only
by default including most of /etc, like /etc/passwd.
For that we need the built-in capability of managing users through the extrausers db,
which Ubuntu Touch uses like Ubuntu Core did previously.

For Ubuntu Touch, in addition to triggering OnClassic being true, introduce an OnTouch
property to easily determine the running system being Ubuntu Touch as determined by
reading /etc/os-release and inspecting for VARIANT_ID=="touch".

fredldotme committed:
daemon: overlord: Enable extrausers on Ubuntu Touch

This allows more sophisticated Snaps, like CUPS, to install and run properly.

fredldotme committed:
interfaces: Fix docker_support AppArmor rules on Ubuntu Touch

Ubuntu Touch like Ubuntu Core 16 might require some additional system-data paths to be
accessible, so add this when running OnTouch.

Additionally, fix installation of the Docker Snap by avoiding conflicting AppArmor
expressions. This results in a Docker environment that cannot start devmode Snaps
but this is not a requirement for Ubuntu Touch, hence proceed using proposed future
AppArmor rules for this.

fredldotme committed:
interfaces/builtin/opengl: Add support for libhybris setups

For libhybris/Halium distributions to support graphics acceleration we need to
allow loading an unknown set of Android libraries which, due to their proprietary
nature, we don't know the exact dependency chain of beforehand. Luckily we
have the OnTouch property to only allow these permissions to be applied on
Ubuntu Touch systems where this might be needed. snap-confine already prepares
the environment appropriately for us, we just need to punch holes into the
security policy of the OpenGL-consuming Snap for an app to consume them.

This allows loading proprietary Android OpenGL ES drivers on Ubuntu Touch.

fredldotme committed:
cmd/snap-confine: Make checks for Halium system more rigid

Instead of just the /system/build.prop file, check for some prominently required
mountpoints, files and symlinks on Halium systems, including whether they actually have
binder devices available in /dev. This should narrow down /android mount setup
and binder device permission setup to just Ubuntu Touch on Halium.

fredldotme committed:
cmd/snap-confine: Run clang-format on some files

fredldotme committed:
cmd/snap-confine: Final cleanup

- Remove creation of the empty hostfs EXTRA_LIB directory, we use hybris-2404 now
- Correct comment about mounting /android into the Snap environment

fredldotme committed:
cmd/snap-confine: Replace hybris' mount checks with regular stat()'s

We don't have access to /proc/mounts and I am uncertain whether shoveling it available
would open another can of worms.

fredldotme committed:
interfaces: Add test for Ubuntu Touch with Halium

Android-based devices on Ubuntu Touch require Halium. Ensure it
and libhybris work and get a usable Snap environment.

fredldotme committed:
cmd/snap-confine: Another clang-format

fredldotme committed:
Revert "cmd/snap-confine: Replace hybris' mount checks with regular stat()'s"

This reverts commit b6fba16e4eeff608a1ad75e6f74862d536e84c96.

fredldotme committed:
cmd: Allow snap-confine read access to /proc/mounts

Required for mount-hybris-support to determine a valid Halium environment.

fredldotme committed:
interfaces/opengl: Strip Halium requirements to bare minimum and comment

fredldotme committed:
interfaces/camera: Add AppArmor snippet for Halium with minimal privileges
